### PR TITLE
Improve the link preview accessibility and labels.

### LINKED ISSUE #60901: LinkControl: Improve the link preview and Copy link aria labels
### Description

The LinkControl has a couple ARIA labels that can be improved.

1
The link preview can show a 'rich data' preview of the link destination page _or_ the link text and url. That's important information that needs to be understandable for all users. Visually, it could be improved with some _visible_ title to explain what it is. Assuming a design change is not desired (although I'd recommend it) the semantics needs to be improved.

In fact, the whole 'preview' _and the buttons_ are wrapped in a `div` element with no ARIA role and [an aria label that says 'Currently selected'](https://github.com/WordPress/gutenberg/blob/e33cb8882134317f7b722011f5d2f39ecc4cecdc/packages/block-editor/src/components/link-control/link-preview.js#L82).
I'm not sure what the reasoning about using such a label was, but it's less than ideal. What 'Currently selected' means in this context?

For screen reader users, this section of information needs to be explained in a meaningful way. 'Currently selected' doesn' tell users anything useful. 

Screenshot of Safari and VoiceOver announcing this section of content as 'Currently selected'. Note that the screen reader also announces 'group' because it attempts to assign the most generic ARIA role to a labelled `div` element with no explicit role set.

Screenshot 2024-04-17 at 12 39 43

2
The copy button label (and tooltip) should not contain the link URL.
- Usign such a label makes this button hardily usable for speech recognition software users.
- It's inconsistend with other Copy buttons.
- The aria-label should never contain _values_ or _states_. Instead, it is meant to provide an accessible name for the control, to explain what that control does.

Screenshot:

Screenshot 2024-04-17 at 12 19 28 2

### Step-by-step reproduction instructions

- Use a screen reader or explore the link preview by using your browser dev tools inspcctor.
- Observe the link preview is wrapped in a div with an aria-label `Currently selected` and it is announced as such.
- Hover or focus the Copy button.
- Observe the tooltip (which is based on the aria-label) contains the URL value.

### Screenshots, screen recording, code snippet

_No response_

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by afercia**
Fixes https://github.com/WordPress/gutenberg/issues/60901

## What?

- The LinkControl 'link preview' popover should use some more meaningful labeling than 'Currently selected'.
- The Copy link button label should not contain the URL value.
- Buttons must not show tooltips when the  'Show button text labels' preference is enabled, as the tooltip just repeats the visibel text.

## Why?

- Meaningful, simple, understandable, labeling is key for usability and accessibility.
- Tooltips are pointless when their text is already visible within the control.

## How?

- The whole popover wrapper has now an ARIA role 'group' with an aria-label 'Manage link'.
- The link preview is now wrapped in an ARIA role 'figure' with an aria-label 'Link information'.
- The Copy link aria-label and tooltip don't contain the URL value any longer.
- The tooltip are no longer displayed when the  'Show button text labels' preference is enabled

I'd appreciate some specific accessibility feedback about the roles and labels.
Cc @joedolson @alexstine 
- Is a role 'group' labeled 'Manage link' meaningful enough? Any suggestion for a better label?
- Is a role 'figure' labeled 'Link information' meaningful enough? Any suggestion for a better label?

## Testing Instructions

- Add a link to some text in a paragraph.
- Once hte link has been added, click or select the link to make the popover appear.
- For keyboard users: just press hte Tab key to move to the popover.
- Use a screen reader or inspect the popover source with your browser dev tools inspector.
- Observe the popover content is wrapped within a `div` element with role `group` and label `Manage link`.
- Observe the link preview within the popover is wrapped within a `span` element with role `figure` and label `Link information`.
- Go to the editor top bar > Options > Preferences > Accessibility, and enable the 'Show button text labels' preference.
- Click the link in the paragraph to open the link popover again.
- Hover or rocus the three buttons within the popover and observe they have visible text and no tooltip.

### CODE REVIEW

File: packages/block-editor/src/components/link-control/link-preview.js
Comment (getdave) [inline packages/block-editor/src/components/link-control/link-preview.js:166]: This was added because users specifically wanted a way to view the full URL without having to edit it. 

For sighted users the URL is clipped when it overflows and thus users requested a way to see the full URL, even if that requires them hovering over a button.

What's the rationale for removing this? Perhaps it's this:

> Tooltips are pointless when their text is already visible within the control.

Which seems true until you realise that the full URL is visually clipped. We don't want to undo that visual clipping (it's been iterated on _several_ times) and so this was a good way to retain UX for sighted users.

File: packages/block-editor/src/components/link-control/link-preview.js
Comment (getdave) [inline packages/block-editor/src/components/link-control/link-preview.js:100]: Can you help me to understand why you feel the word manage is more appropriate here?

This is a preview of the currently selected link so that's why it's named as such.

File: packages/block-editor/src/components/link-control/link-preview.js
Comment (getdave) [inline packages/block-editor/src/components/link-control/link-preview.js:178]: Thanks for spotting this.

**Review by getdave — COMMENTED** [review]
Thank you for iterating on the a11y and UX here. 

I have a few questions just so I can feel comfortable with some of the changes.

File: packages/block-editor/src/components/link-control/link-preview.js
Comment (afercia) [reply packages/block-editor/src/components/link-control/link-preview.js:166]: It's not only a visual thing. You are missing the point that before being the tooltip, this is the aria-label of the button.
The aria-label provides the accessible name of a control. A name control is not the place for values or states. It just has to give the control a name that tells users what the control does.

This is basically the equivalent of the button text, would you ever provide users with a HTML button like the following one? 

Screenshot 2024-04-23 at 09 26 51

If we want to provide users the ability to see the full URL, then the tooltip is not the right place for that. I also want to mention that these are simple, basic, accessibility principles we've been repeating sinc eyears in this project but it seems designers still don't get the point.
We can't use tooltips this way. Tooltips were introduced in Gutenberg with the only specific purpose to visually expose an icon button name. They must not be used for values, states, or descriptions.

File: packages/block-editor/src/components/link-control/link-preview.js
Comment (afercia) [reply packages/block-editor/src/components/link-control/link-preview.js:100]: @getdave, did you read the issue before asking your question? I think I already explained there the reasoning.

File: packages/block-editor/src/components/link-control/link-preview.js
Comment (afercia) [reply packages/block-editor/src/components/link-control/link-preview.js:178]: It is extremely annoying that contributors to this project keep missing this. That also proves there is a lack of manual testing and no one tests the UI with the 'Show button text labels' preference enabled, which is less than ideal.

File: packages/block-editor/src/components/link-control/link-preview.js
Comment (getdave) [reply packages/block-editor/src/components/link-control/link-preview.js:100]: Yes I read it.

However I did miss the fact that you'd wrapped the _entire_ `LinkPreview` component in a `Manage link` group. That was an oversight on my part which I was happy to correct.

File: packages/block-editor/src/components/link-control/link-preview.js
Comment (getdave) [reply packages/block-editor/src/components/link-control/link-preview.js:166]: What I've tried to do is provide you with context you are missing regarding a UX requirement of being able to perceive the entire link. If that context doesn't currently marry with best practices then it would be nice if we could work together collaboratively and constructively to devise a more suitable solution for the benefit of users.

Do you have any suggestions for how we could achieve this?

> I also want to mention that these are simple, basic, accessibility principles we've been repeating sinc eyears in this project 

I appreciate your frustration regarding accessibility principles. I looked at the component docs and noted the lack of documentation around this component and how it should be used:

https://wordpress.github.io/gutenberg/?path=/docs/components-tooltip--docs

Given that this something you find yourself repeating, I think we could work together to improve the documentation and add specific notes on accessibility linking out to resources where appropriate.

I'll hold my hand up and accept that I could be more informed about the specifics of a11y regarding tooltips, but I strongly believe that if we want consumers (especially 3rd party, non-Core team developers) to follow best practices we should be looking at providing _documentation_ alongside the component. 

That requires a collaborative effort and I'd be more than happy to work alongside you on that if you are open to it. If not then I will do it myself.

> ...but it seems designers still don't get the point.

I'm not a designer but I think it's disingenuous to paint designers in this light. Mistakes happen, it's the will to correct and improve both the code and one's own knowledge that is key. Providing documentation is also a great way to share the knowledge of individual contributors and make it more readily available to others.

File: packages/block-editor/src/components/link-control/link-preview.js
Comment (getdave) [reply packages/block-editor/src/components/link-control/link-preview.js:178]: > It is extremely annoying that contributors to this project keep missing this. 

Again the documentation makes no mention of this. We don't have to make it wordpress specific but mentioning when this property should / shouldn't be applied would help avoid contributors needing to monitor for further reoccurrences.

https://wordpress.github.io/gutenberg/?path=/docs/components-button--docs

> That also proves there is a lack of manual testing 

I don't think it's helpful to draw such conclusions from a single PR review instance.

In this case I didn't perform manual testing as it was a first pass review. Happy to do so once the issue of preserving the ability to perceive the full URL from the preview has been addressed.

> ...and no one tests the UI with the 'Show button text labels' preference enabled

I find it's usually best to avoid generalisations such as this. Indeed, whilst I won't name them, I can think of _many_ contributors (including designers) who do this on a regular basis.

File: packages/block-editor/src/components/link-control/test/index.js
Comment (getdave) [inline packages/block-editor/src/components/link-control/test/index.js]: We should prefer the `getByRole` selector.  This is [documented in the guide for Playwright](https://github.com/WordPress/gutenberg/blob/73a17b0d3dc9e08aaeb84bf03c9398d61a327a8c/docs/contributors/code/e2e/README.md#use-accessible-selectors) and we should aim to abide by this when writing tests.
```suggestion
			const linkPreview = screen.getByRole( 'group', {
				name: 'Manage link',
			} );
```

There are other occurrences within this file that also need fixing to utilise the new `role`s you've added.

File: packages/block-editor/src/components/link-control/link-preview.js
Comment (getdave) [reply packages/block-editor/src/components/link-control/link-preview.js:166]: For avoidance of confusion here's an example of where being able to see the entire URL is needed. Is this linking to the correct PDF? How do I know? 

Screen Shot 2024-04-23 at 10 56 40

Again, I'm not advocating we continue to use a pattern that provides a poor experience for users of assistive tech but I feel we owe it to our users to find a solution that accounts for this requirement before we remove it from the UI.

File: packages/block-editor/src/components/link-control/test/index.js
Comment (afercia) [reply packages/block-editor/src/components/link-control/test/index.js]: I'd agree usint the role selector is better, but this is already existing code, I was just changing the label value.

**Comment by afercia** [comment]
Re: Documentation: this is more about the `label` prop of the Button component. Reference:

https://github.com/WordPress/gutenberg/blob/1f6fec6575ca4e9bcd64822b533df9f3057b9ba1/packages/components/src/button/README.md#label-string

The documentation clearly states the `label` prop sets the `aria-label` of the component. I expect developers in this project to know what an aria-label attribute is and how it should be used. If they don't know what an aria-label is, I escpet them to read the W3C documentation about it and learn from there. References:
https://www.w3.org/TR/wai-aria-1.2/#aria-label
https://www.w3.org/TR/wai-aria-1.2/#namecalculation

## Quick history of Tooltips in Gutenberg

I'll try to provide some more context about the original implementation of Tooltips, hoping it's a good base for some better documentation. 

Suggested in the initial accessibility recommendations for Gutenberg on Mar 21, 2017. See point 8. on this issue:
https://github.com/WordPress/gutenberg/issues/297
> The icons (such as for headings and arrows and the like) will need some sort of visual explanation. I would not use a title attribute, but instead look at a programmatic tooltip that can be enabled/disabled as a user setting.

Noting they were specifically requested for _icon buttons_.

Issue regarding implementing Tooltips, created on Apr 27, 2017:

**Block controls: Show tooltip on hover, keyboard focus**
https://github.com/WordPress/gutenberg/issues/531

First implementation of Tooltips was made in this PR on Aug 8, 2017:

**Components: Add Tooltip component, display button aria-label tooltips**
https://github.com/WordPress/gutenberg/pull/2293

Noting the PR title itself mentions the purpose to 'display button aria-label'.

**Comment by afercia** [comment]
> I don't think it's helpful to draw such conclusions from a single PR review instance.

I'm sorry if that sounded unfair. However, it's not just a single case. In the last years I've been observing countless changes to the editor UI that didn't take into consideration the 'Show button text labels' preference at all, thuys introducing breakages in the UI. It's clearly an editor feature that is under-tested and often forgot in many cases. I had to submit a few PRs in the past to fix those cases but unfortunately that happens way after the breakage was merged and released.

**Comment by afercia** [comment]
To better clarify the current situation on trunk. The whole preview popover is wrapped within an element labeled `Currently selected`, which doesn't make much sense. It really doesn't explain users what this popover content is about.

To illustrate visually: all the UI within the highlighted area in the screenshot below is labeled `Currently selected`:

Screenshot 2024-04-24 at 12 52 29

The UI is actually made of multiple things:
- The 'link preview', which I would more correctly call 'link _destination_ preview'.
- 3 buttons to Edit, Remove, Copy

Semantically, this section of content is a `div` element with an `aria-label` attribute. The div has no specific ARIA role set so it is exposed by browser as `role=generic`:

Screenshot 2024-04-24 at 12 58 16

Setting an `aria-label` on an element with no role doesn't make much sense in the first place. Some screen readers try to assign the most generic role `group` because they see an aria-label but that's a screen-reader internal implementation.

The solution is to either set an explicit ARIA role or remove the aria-label. However, removing the aria-label would let users without any clue what this popover is about.

I tend to think an explicit `role=group` is the best thing to do here. You can think a `role=group` as kind of a more generic semantics of 'fieldset' HTML element. It aims to group together logically related controls.

About the value of the aria-label, I'm not sure 'Manage' is the best term to use. Considering the diverse nature of the popover content I can't think of a better name.

As a last note: the 'link preview' may or may not be an actual 'preview'. First of all, it doesn't preview the 'link', rather it previews the 'destination resource' of the link, in most of the cases a web page. However, when the destination can't be 
fetched, what the UI actually shows isn't really a 'preview'. It's just the link text. E.g.:

Screenshot 2024-04-24 at 12 52 09

**Comment by afercia** [comment]
Re: the design issue and hte need for users to see the full URL. Long URLs have alwasy been a problem. There's no optimal solution for that. However, usitn the tooltip is far from ideal:
- It breaks semantics and accessibility, because the URL is part of the aria-label.
- Visuallly, very long URLs in the tooltip aren't great either. See the two example screenshots below.

Long URL shown within the Copy button tooltip:

Screenshot 2024-04-24 at 14 02 47

Very very long URL shown within the Copy button tooltip:

Screenshot 2024-04-24 at 14 29 55

If the full URL is considered important information to provide to users (and I'd agree it is) then it should be shown with visibel text in the UI. 

Howeer, very long URLs are a problem. It's not a new problem though. It has been discussed at length and addressed previously in Core, for hte wp-link toolbar for TinyMCE. 

While truncating text isn't ideal, at some point very long URLs need to be truncated. wplink tries to do that in a smart way and I'd suggest to just copy what WordPress has been doing for years to display long URLs in the link 'preview'.

The implementation is here: https://github.com/WordPress/wordpress-develop/blob/d898a25d1a37f493b7050b263cb9c5b29565792a/src/js/_enqueues/vendor/tinymce/plugins/wplink/plugin.js#L11-L51

- First of all, I'm not sure why this Gutenberg popover width is limited to a `max-width: 350px`. This popover potentially contains very long text. Both the link _text_  and hte link URL may be very long. The max-width should be increased a little.
- For long URLs:
  - The Classic wplink strips out anything that starts with `?` or `#` first.
  - Then it strips out any `index.html` from the end of the URL, if any.
  - Then it strips out any ending forward slash `/`. 
  - It then preserves the URL if it's just a fragment identifier e.g. `#some-id-on-the-page`.
  - Then, the 'smart' part: when the URL is longer than 40 characters, it tries to replace any 'in-between' portion of the URL with an ellipsis in order to show only the first and last part of the URLs, assuming those are the most important parts to display.

Besides this JavaScript part, [the CSS truncates the URL anyways](https://github.com/WordPress/wordpress-develop/blob/d898a25d1a37f493b7050b263cb9c5b29565792a/src/wp-includes/css/editor.css#L1719-L1725), this is the CSS used on Classic editor for the `wp-link-preview` wrapper:
```
    max-width: 694px;
    overflow: hidden;
    text-overflow: ellipsis;
```

Since this hae been the implementation in Core for years and it worked reasonably, I'd suggest to just use a similar implementation. As said, there's no perfect solution to display very long URLs.

I will create a separate issue.

Screenshots from Classic editor link 'preview':

Screenshot 2024-04-24 at 14 05 15

Screenshot 2024-04-24 at 14 10 02

**Comment by afercia** [comment]
I created a new issue https://github.com/WordPress/gutenberg/issues/61158 to explore a new way to display long URLs.

I'd appreciate a new review of this PR, thanks everyone.

**Comment by afercia** [comment]
@WordPress/gutenberg-core I'd appreciate a new review, when you have a chance.
For a recap, I suggest start reading from https://github.com/WordPress/gutenberg/pull/60908#issuecomment-2074703836

File: packages/block-editor/src/components/link-control/link-preview.js
Comment (getdave) [reply packages/block-editor/src/components/link-control/link-preview.js:166]: I'd like to reiterate my concerns above regarding the tooltip providing important UX for sighted users.

I've re-read the thread and acknowledhge that `Copy link: {{FULL URL HERE}}` is not helpful as an accessible `name`.

<details>
<summary>Screenshot of the long name for context</summary>
Screen Shot 2024-10-24 at 12 32 43
</details>

That said whilst fixing this for one set of users, we should avoid introducing a regression for another set.

Therefore I wonder if it's possible to have a tooltip attached to the button which isn't used as the `name` for the button but which is perceivable by all users _if they choose_? What a11y roles or semantics could be used to achieve this?

File: packages/block-editor/src/components/link-control/link-preview.js
Comment (getdave) [reply packages/block-editor/src/components/link-control/link-preview.js:178]: I'd recommend we be proactive here and raise another PR to update the documentation for this prop to add some guidance about it's effective usage. We could explain the a11y best practices that surround this and help to guide future contributors.

Any other components that make use of this prop could also be covered. 

I wonder what @ciampo thinks about improving the guidance around relying on icons/tooltips in the component library?

File: packages/block-editor/src/components/link-control/link-preview.js
Comment (getdave) [inline packages/block-editor/src/components/link-control/link-preview.js:100]: The `Currently selected` is quite old code now and I'm really pleased to get input on improving this 👍

File: packages/block-editor/src/components/link-control/link-preview.js
Comment (getdave) [inline packages/block-editor/src/components/link-control/link-preview.js:117]: This seems like a good option 👍

File: packages/block-editor/src/components/link-control/link-preview.js
Comment (getdave) [reply packages/block-editor/src/components/link-control/link-preview.js:166]: I see you are exploring in https://github.com/WordPress/gutenberg/issues/61158. I'll check that too.

Comment (getdave) [reply packages/block-editor/src/components/link-control/link-preview.js:166]: I've re-tested the implementation on this PR and I've changed my opinion. Having a very long URL in a tooltip doesn't offer a huge amount of value. It is also detrimental to the UX for a large number of users.

With the advancements to the Link UI in WordPress 6.5 it's now much easier to just click "Edit" and see the full URL in the form input field. Prior to those changes it was necessary to click several buttons to achieve the same thing.

So overall I'd be in favour of this change.

**Review by getdave — APPROVED** [review]
This seems to be working well and improves the accessibility of an important piece of the Editor interface.

I did have concerns regarding removing the ability to hover and see the full URL in a tooltip as this was added specifically to aid with sighted users who regularly add long URLs (e.g. PDF files...etc).

However with recent changes to the UI you can easily hit edit and check the URL from there. It's not perfect but I think the trade off is worth it to improve the overall accessibility of the control.

File: packages/block-editor/src/components/link-control/link-preview.js
Comment (ciampo) [reply packages/block-editor/src/components/link-control/link-preview.js:178]: Better documentation is always a win!

I should also mention that it was recently agreed (see https://github.com/WordPress/gutenberg/pull/65989 and https://github.com/WordPress/gutenberg/pull/66021) that `Tooltip`s should be purely visual/decorative, and therefore should not be intended as a tool to add accessible descriptions to their anchors. 

The anchor should be self-sufficient when it comes to their accessible label / description — ie. they should either have accessible content, or use the `aria-label` / `aria-labelledby` / `aria-describedby` attributes accordingly.

I can look into updating `Button`'s and `Tooltip`'s documentation accordingly.

**Comment by afercia** [comment]
> I should also mention that it was recently agreed (see https://github.com/WordPress/gutenberg/pull/65989 and https://github.com/WordPress/gutenberg/pull/66021) that Tooltips should be purely visual/decorative, and therefore should not be intended as a tool to add accessible descriptions to their anchors.

> The anchor should be self-sufficient when it comes to their accessible label / description — ie. they should either have accessible content, or use the aria-label / aria-labelledby / aria-describedby attributes accordingly.

@ciampo can you please clarify, specifically `Tooltips should be purely visual/decorative`? While I do agree they should not be used for _descriptions_, it has been mentioned several times that for icon buttons the tooltip needs to visually expose the accessible name (the 'label'). As such, _it does_ relate to accessibility.

**Comment by afercia** [comment]
> I've re-tested the implementation on this PR and I've changed my opinion. Having a very long URL in a tooltip doesn't offer a huge amount of value. It is also detrimental to the UX for a large number of users.

Thanks @getdave for your review. Yes, that was one of my points.

> I see you are exploring in #61158. I'll check that too.

Yes and I'd encourage everyone to explore the broader issue of 'how to fully display long values' across the whole editor. It's a broader issue and, in my opinion, some better approach should be explored for other cases as well. Long values need space. It's not ideal to display these values in parts of the UI that have a limited width by their own nature.

**Comment by ciampo** [comment]
> @ciampo can you please clarify, specifically `Tooltips should be purely visual/decorative`? While I do agree they should not be used for _descriptions_, it has been mentioned several times that for icon buttons the tooltip needs to visually expose the accessible name (the 'label'). As such, _it does_ relate to accessibility.

I believe we're agreeing on this one.

By "Tooltips should be purely visual/decorative" I meant that tooltips should not provide accessible labels and/or descriptions to the browser (especially given that the `Tooltip` component renders conditionally), and that anchors should be self-sufficient in providing those. Tooltips still play an important role to sighted users, visually exposing such information (hence the "visual" word in my sentence). Also, the previous sentence continues with "...and therefore should not be intended as a tool to add accessible descriptions to their anchors.".

**Comment by afercia** [comment]
Thanks for clarifying @ciampo. Still, I'd like to make clear to all contributors who may read this conversation in the future, that icon buttons _must_ visually expose via a tooltip the accessible name provided via their `aria-label` attribute. This is essential for accessibility e.g. for voice control / speech recognition software users, for any user who may not recognize the meaning of icons, etc.

**Pull request merged**
